File: src/python/problems/unique_occurrences_test.py
# https://leetcode.com/problems/unique-number-of-occurrences

from core.problem_base import *

class UniqueOccurrences(ProblemBase):
    def Solution(self, arr: list[int]) -> bool:
        map = {}
        for i in range(len(arr)):
            if arr[i] not in map:
                map[arr[i]] = 1
            else:
                map[arr[i]] += 1
        
        unique = set(map.values())

        return len(unique) == len(map.values())
        
if __name__ == '__main__':
    TestGen(UniqueOccurrences) \
        .Add(lambda tc: tc.Param("arr", [1,2,2,1,1,3]).Result(True)) \
        .Add(lambda tc: tc.Param("arr", [1,2]).Result(False)) \
        .Add(lambda tc: tc.Param("arr", [-3,0,1,-3,1,1,1,-3,10,0]).Result(True)) \
        .Run()

File: src/go/problems/unique_occurrences_test.go
// https://leetcode.com/problems/unique-number-of-occurrences/

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type UniqueOccurrences struct{}

func TestUniqueOccurrences(t *testing.T) {
	gen := core.TestSuite[UniqueOccurrences]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1,2,2,1,1,3}).Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1,2}).Result(false)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{-3,0,1,-3,1,1,1,-3,10,0}).Result(true)
	}).Run(t)
}

func (UniqueOccurrences) Solution(arr []int) bool {
	occ := map[int]int{}
	for _,v := range arr {
		occ[v]++
	}

	unique := map[int]struct{}{}
	for _,v := range occ {
		unique[v] = struct{}{}
	}

	return len(occ) == len(unique)
}
